core/src/main/java/io/questdb/cairo/TxnScoreboardV2.java

```diff
@@ -0,0 +1,262 @@
+
+        long min = Long.MAX_VALUE;
+        for (int i = 0; i < entryScanCount; i++) {
+            long lockedTxn = Unsafe.getUnsafe().getLongVolatile(null, entriesMem + (long) i * Long.BYTES);
+            if (lockedTxn > UNLOCKED) {
+                min = Math.min(min, lockedTxn);
+            }
+        }
+        return min == Long.MAX_VALUE ? UNLOCKED : min;
+    }
+
+    @Override
+    public TableToken getTableToken() {
+        return tableToken;
+    }
+
+    @Override
+    public boolean hasEarlierTxnLocks(long txn) {
+        // Push max txn to the latest, but don't stop checking if it's not the max.
+        updateMax(txn);
+
+        if (getActiveReaderCount() == 0) {
+            return false;
+        }
+
+        long scanCount = readScanCount();
+        for (int i = 0; i < scanCount; i++) {
+            long lockedTxn = Unsafe.getUnsafe().getLongVolatile(null, entriesMem + (long) i * Long.BYTES);
+            if (lockedTxn > UNLOCKED && lockedTxn < txn) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public boolean incrementTxn(int id, long txn) {
+        long internalId = toInternalId(id);
+        assert internalId < entryScanCount;
+        if (Unsafe.getUnsafe().compareAndSwapLong(null, entriesMem + internalId * Long.BYTES, UNLOCKED, txn)) {
+            updateMaxReaderId(internalId);
+            incrementActiveReaderCount();
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean isMax(long txn) {
+        return txn >= getMax();
+    }
+
+    @Override
+    public boolean isRangeAvailable(long fromTxn, long toTxn) {
+        if (getActiveReaderCount() == 0) {
+            return true;
+        }
+
+        long scanCount = readScanCount();
+        for (int i = 0; i < scanCount; i++) {
+            long lockedTxn = Unsafe.getUnsafe().getLongVolatile(null, entriesMem + (long) i * Long.BYTES);
+            if (lockedTxn > UNLOCKED && lockedTxn >= fromTxn && lockedTxn < toTxn) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public boolean isTxnAvailable(long txn) {
+        if (getActiveReaderCount() == 0) {
+            return true;
+        }
+
+        long scanCount = readScanCount();
+        for (int i = 0; i < scanCount; i++) {
+            long lockedTxn = Unsafe.getUnsafe().getLongVolatile(null, entriesMem + (long) i * Long.BYTES);
+            if (lockedTxn == txn) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public long releaseTxn(int id, long txn) {
+        long internalId = toInternalId(id);
+        assert internalId < entryScanCount;
+        Unsafe.getUnsafe().putLongVolatile(null, entriesMem + internalId * Long.BYTES, UNLOCKED);
+        // It's too expensive to count how many readers are left for the txn, caller will have to do additional checks.
```

Comment:
nit: it might be a good idea to change this method to `void` and make additional checks unconditionally in `TableReader#goPassive()` when deciding whether to schedule O3 partitions.

Reply:
It's for compatibility with V1 as a backup plan if there will be problems with V2 and we need to switch back. We can refactor it in next releases when we remove V1 implementation.

Reply:
I'm not sure if removing V1 any time soon is a good idea considering that it provides the only way to go multi-process. Aligning the APIs and making them less inconsistent wouldn't hurt in any case, especially considering that most of such changes would be V2-driven.

Reply:
We don't support multi-process now

Reply:
Yet, some folks use it.